zokrates_cli/src/ops/mpc/contribute.rs

```diff
@@ -0,0 +1,115 @@
+use crate::constants::MPC_DEFAULT_PATH;
+use clap::{App, Arg, ArgMatches, SubCommand};
+use phase2::parameters::MPCParameters;
+use std::fs::File;
+use std::io::{BufReader, BufWriter};
+use std::path::Path;
+
+pub fn subcommand() -> App<'static, 'static> {
+    SubCommand::with_name("contribute")
+        .about("Contributes to a ceremony")
+        .display_order(2)
+        .arg(
+            Arg::with_name("input")
+                .short("i")
+                .long("input")
+                .help("Path of the MPC parameters")
+                .value_name("FILE")
+                .takes_value(true)
+                .required(false)
+                .default_value(MPC_DEFAULT_PATH),
+        )
+        .arg(
+            Arg::with_name("entropy")
+                .short("e")
+                .long("entropy")
+                .help("User provided randomness")
+                .takes_value(true)
+                .required(false),
+        )
+        .arg(
+            Arg::with_name("output")
+                .short("o")
+                .long("output")
+                .help("Path of the output file")
+                .value_name("FILE")
+                .takes_value(true)
+                .required(false)
+                .default_value(MPC_DEFAULT_PATH),
+        )
+}
+
+pub fn exec(sub_matches: &ArgMatches) -> Result<(), String> {
+    let path = Path::new(sub_matches.value_of("input").unwrap());
+    let file =
+        File::open(&path).map_err(|why| format!("Could not open `{}`: {}", path.display(), why))?;
+
+    let reader = BufReader::new(file);
+    let mut params = MPCParameters::read(reader, true)
+        .map_err(|why| format!("Could not read `{}`: {}", path.display(), why))?;
+
+    let entropy = sub_matches.value_of("entropy").unwrap();
+
+    // Create an RNG based on a mixture of system randomness and user provided randomness
+    let mut rng = {
+        use blake2::{Blake2b, Digest};
+        use byteorder::{BigEndian, ReadBytesExt};
+        use rand::chacha::ChaChaRng;
+        use rand::{OsRng, Rng, SeedableRng};
+
+        let h = {
+            let mut system_rng = OsRng::new().unwrap();
+            let mut h = Blake2b::default();
+
+            // Gather 1024 bytes of entropy from the system
+            for _ in 0..1024 {
+                let r: u8 = system_rng.gen();
```

Comment:
The tutorial mentions different types of randomness that can be used here, but we always use the Os?

Reply:
RNG is based on a mixture of system randomness and user-provided randomness

Reply:
https://github.com/Zokrates/ZoKrates/blob/mpc/zokrates_cli/src/ops/mpc/contribute.rs#L71 user-provided randomness is applied here

Reply:
I see! For cryptographic code like this I feel it would be good to have a separate crate and reuse external tools as much as possible, and if not, give links to references.